### ISSUE TITLE
[inductor] Mix-order reduction silently corrupts conv-bias grads (bf16, sm_120) in 2.10.0 — fixed in 2.11 without regression test?

### ISSUE THREAD

**Issue by Euswbnix** [pytorch/pytorch#190796]
### 🐛 Describe the bug

On torch **2.10.0+cu128** (triton 3.6.0, RTX 5090 / sm_120), `torch.compile` (default inductor backend) + `torch.autocast("cuda", torch.bfloat16)` produces **corrupted gradients** for the bias of a `Conv2d` whose output receives a broadcasted per-channel add (the standard diffusion-UNet ResBlock timestep-embedding pattern):

```python
self.conv(x) + self.proj(F.silu(temb))[:, :, None, None]
```

The backward for this pattern is compiled into a fused **mix-order reduction** kernel (`grid_type: 'MixOrderReductionGrid'`, workspace `ws_ptr` argument) that computes the conv-bias sum and the Linear-backward sums in one kernel. On 2.10.0 the resulting `conv.bias.grad` is **wrong on every invocation** — max abs error ~1e37–1e38 vs `aot_eager` on identical inputs — and frequently contains NaN (whenever the garbage memory it reads happens to hold NaN bit patterns). All other grads (conv weight, Linear weight/bias) match `aot_eager` to normal bf16 tolerance, and the forward is unaffected.

We hit this in real training: a 35.7M-param DDPM UNet on CIFAR-10 got NaN grads within 3–4 optimizer steps. (`clip_grad_norm_` then smears the NaN into every other grad, which makes the first-NaN localization misleading — inspect grads before clipping.)

### Minimal repro

Deterministic on our machine: NaN in `conv.bias.grad` on the **first backward call**, same element indices on every run. 10 fresh random batches are attempted because the NaN (unlike the underlying corruption) depends on the contents of recycled allocator memory.

```python
# NaN repro: torch.compile (inductor) + bf16 autocast, RTX 5090 (sm_120), torch 2.10.0+cu128
import torch
import torch.nn as nn
import torch.nn.functional as F

print(torch.__version__, torch.version.cuda, torch.cuda.get_device_name())


class Block(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(512, 256, 3, padding=1)
        self.proj = nn.Linear(512, 256)

    def forward(self, x, temb):
        return self.conv(x) + self.proj(F.silu(temb))[:, :, None, None]


torch.manual_seed(0)
model = Block().cuda()
compiled = torch.compile(model)

for step in range(10):
    x = torch.randn(128, 512, 16, 16, device="cuda")
    temb = torch.randn(128, 512, device="cuda")
    y = torch.randn(128, 256, 16, 16, device="cuda")
    with torch.autocast("cuda", torch.bfloat16):
        loss = F.mse_loss(compiled(x, temb), y)
    model.zero_grad(set_to_none=True)
    loss.backward()
    bad = {n: p.grad for n, p in model.named_parameters()
           if not torch.isfinite(p.grad).all()}
    for n, g in bad.items():
        idx = (~torch.isfinite(g)).nonzero().flatten().tolist()
        print(f"step {step}: {n}.grad has {len(idx)} non-finite element(s) at {idx}")
    if bad:
        raise SystemExit(1)
    print(f"step {step}: grads finite")
print("no NaNs (bug not reproduced)")
```

Output on 2.10.0+cu128:

```
2.10.0+cu128 12.8 NVIDIA GeForce RTX 5090
step 0: conv.bias.grad has 2 non-finite element(s) at [184, 198]
```

### Silent corruption, not just NaN

Comparing `conv.bias.grad` against `backend="aot_eager"` on identical weights/data (6 fixed batches, seed 0): **every batch** is corrupted under inductor, whether or not it happens to contain NaN. All other parameter grads match to bf16 tolerance (~6e-3 rel):

```
batch 0 conv.bias  max_abs_diff=2.472e+38  <-- NON-FINITE (inductor)
batch 1 conv.bias  max_abs_diff=2.658e+38  <-- NON-FINITE (inductor)
batch 2 conv.bias  max_abs_diff=2.260e+37  <-- NON-FINITE (inductor)
batch 3 conv.bias  max_abs_diff=1.296e+37  <-- finite but garbage
batch 4 conv.bias  max_abs_diff=2.712e+38  <-- finite but garbage
batch 5 conv.bias  max_abs_diff=2.492e+37  <-- finite but garbage
```

So the NaN is only the visible tip; the kernel reads garbage memory into the bias-grad accumulation on every call. Grad-clipped training can be silently degraded without ever surfacing a NaN.

### What we ruled out (all verified by execution on the same machine)

| configuration | result |
|---|---|
| inductor + bf16 autocast (default) | **corrupted/NaN** |
| `backend="aot_eager"` + bf16 | clean |
| eager + bf16 | clean |
| inductor + fp32 (no autocast) | clean |
| inductor + bf16, batch 32 / 64 (reduction below mix-order threshold) | clean |
| `torch._inductor.config.triton.persistent_reductions = False` | still corrupted |
| `torch._inductor.config.split_reductions = False` | still corrupted |
| `torch._inductor.config.triton.cooperative_reductions = False` | still corrupted |
| `emulate_precision_casts = True` | still corrupted |
| `mode="max-autotune-no-cudagraphs"` | still corrupted |
| **`TORCHINDUCTOR_MIX_ORDER_REDUCTION=0`** | **clean** (workaround) |

In the original UNet the failure also occurred with the SDPA math backend and with dropout 0, and under `channels_last`; attention turned out to be entirely irrelevant (the minimal repro has none).

### Generated-kernel forensics

The backward compiles to (TORCH_LOGS=output_code, 2.10.0):

```python
workspace_0 = empty_strided_cuda((524288, ), (1, ), torch.float32)  # uninitialized
triton_per_fused__to_copy_convolution_backward_sum_0.run(
    tangents_1, buf1, workspace_0, 32768, 256, stream=stream0)
buf6 = workspace_0[0 : 2048 * 256].view(2048, 256).sum(dim=0)       # bias grad
```

with the kernel (`inductor_meta` includes `'grid_type': 'MixOrderReductionGrid'`, `'RSPLIT_SIZE': 16`):

```python
@triton.jit
def triton_per_fused__to_copy_convolution_backward_sum_0(in_ptr0, out_ptr1, ws_ptr, xnumel, r0_numel, XBLOCK: tl.constexpr, RSPLIT_SIZE: tl.constexpr, NUM_STAGES: tl.constexpr):
    xnumel = 32768
    r0_numel = 256
    ...
    xoffset = tl.program_id(0) * RSPLIT_SIZE
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    ...
    accum0 = tl.full([R0_BLOCK], 0, tl.float32)[None, :]
    split_size = min(RSPLIT_SIZE, xnumel - xoffset)
    for _ in tl.range(0, split_size, XBLOCK, num_stages=NUM_STAGES):
        x0 = xindex
        x4 = xindex // 256
        x5 = (xindex % 256)
        xindex += XBLOCK
        tmp0 = tl.load(in_ptr0 + (r0_1 + 256*x0), None)              # unmasked
        tmp6 = tl.load(in_ptr0 + (x5 + 256*r0_1 + 65536*x4), None)   # unmasked
        ...
        tl.store(out_ptr1 + (x0), tmp5, None)
        tmp7 = tl.sum(tmp6, 0)
        accum0 = accum0 + tmp7
    tl.store(ws_ptr + (tl.program_id(0) + 0 * tl.num_programs(0)) * r0_numel + r0_index, accum0, r0_mask)
```

Observations (hypotheses, not verified root cause):
- All loads/stores are unmasked (`mask=None`) and the loop advances `xindex` manually while being software-pipelined (`num_stages=NUM_STAGES`). In 2.10.0, `NUM_STAGES = min(max(num_iters // 4, 1), 3)` unconditionally (`triton_heuristics.py`); pipelined prefetch of a manually-advanced unmasked load can read out of bounds on the final iteration(s).
- The fp32 workspace is allocated with `empty_strided_cuda` (uninitialized) and the wrapper unconditionally sums 2048 rows of it.
- The corrupted values (~1e38, occasional NaN) look exactly like recycled-allocator garbage entering the accumulation. In configs with little prior allocator activity, step 0 is often clean (pristine pages) and corruption appears once the caching allocator starts recycling; in the repro above, compile/warmup allocations have already churned the pool before the first backward, so it fails immediately — both observations are consistent.

### Version matrix (same machine, same script)

| torch | triton | result |
|---|---|---|
| 2.10.0+cu128 | 3.6.0 | **corrupted/NaN** |
| 2.11.0+cu128 | 3.7.x | clean |
| 2.12.1+cu129 | 3.7.x | clean |
| 2.13.0+cu130 | 3.7.1 | clean |
| 2.14.0.dev20260722+cu130 (nightly) | pinned | clean |

Notably, 2.13.0 and nightly still emit a near-identical mix-order kernel (same unmasked loads, same uninitialized workspace + wrapper tail-sum) for this pattern — the generated Triton source differs only in eviction-policy hints. So whatever fixed this between 2.10 and 2.11 appears to live in the launch-config heuristics (e.g. the `NUM_STAGES`/`XBLOCK` selection, which 2.13 gates via `mix_order_reduction_allow_multi_stages`) or in the bundled Triton compiler, not in the emitted kernel body.

### Asks

1. Can someone identify which change fixed this between 2.10.0 and 2.11.0, and confirm it was an intentional correctness fix rather than a heuristic shift that happens to avoid the bad configuration on this shape?
2. Since the emitted kernel body is unchanged in 2.13/nightly, a regression test for this exact pattern (conv-bias grad + broadcast-add backward, bf16, reduction ≥ 32k elements) would guard against the bad launch configuration coming back.
3. Worth backporting if a 2.10.x patch release happens; otherwise, a searchable record for 2.10 users (`TORCHINDUCTOR_MIX_ORDER_REDUCTION=0` is a clean workaround).

Related (different defects in the same subsystem): #186241 (TMA descriptor stale xoffset in mix-order loop), #185534, #185814.

### Versions

```
PyTorch version: 2.10.0+cu128
CUDA used to build PyTorch: 12.8
OS: Ubuntu 24.04.4 LTS (x86_64)
Python version: 3.10.18 (64-bit runtime)
GPU: NVIDIA GeForce RTX 5090 (sm_120)
Nvidia driver version: 590.48.01
triton==3.6.0
```

cc @chauhang @penguinwu

**Comment by zou3519** [comment]
I cannot repro this on nightly. Also, the issue says this is fixed in 2.11, so I'm inclined to close this issue.

**Comment by drisspg** [comment]
Closing as fixed
